### PR TITLE
add ADE20K dataset

### PR DESCRIPTION
**Pull request by walkerlala**
This PR provide scripts/documents for downloading/converting the ADE20K dataset and training deeplabv3 on it. One particular thing to note about is the `exclude_list` variable in `utils/train_utils.py`. I think it is OK to do this, but it might hurt performance of other models.

### CODE REVIEW

File: research/deeplab/datasets/build_ade20k_data.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  Args:
+    dataset_split: dataset split (e.g., train, val)
+    dataset_dir: dir in which the dataset locates
+    dataset_label_dir: dir in which the annotations locates
+
+  Raises:
+    RuntimeError: If loaded image and label have different shape.
+  """
+
+  img_names = glob.glob(os.path.join(dataset_dir, '*.jpg'))
```
Comment (YknZhu) [inline research/deeplab/datasets/build_ade20k_data.py]: maybe tf.gfile.Glob(...)?

File: research/deeplab/datasets/build_ade20k_data.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  seg_names = []
+  for f in img_names:
+    # get the filename without the extension
+    basename = os.path.basename(f).split(".")[0]
+    # cover its corresponding *_seg.png
+    seg = os.path.join(dataset_label_dir, basename+'.png')
+    seg_names.append(seg)
+
+  num_images = len(img_names)
+  num_per_shard = int(math.ceil(num_images) / float(_NUM_SHARDS))
```
Comment (YknZhu) [inline research/deeplab/datasets/build_ade20k_data.py]: math.ceil(num_images / float(_NUM_SHARDS)) ?

File: research/deeplab/datasets/download_and_convert_ade20k.sh [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+#         + images
+#           + training
+#           + validation
+
+# Exit immediately if a command exits with a non-zero status.
+set -e
+
+CURRENT_DIR=$(pwd)
+WORK_DIR="./ADE20K"
+mkdir -p ${WORK_DIR}
```
Comment (YknZhu) [inline research/deeplab/datasets/download_and_convert_ade20k.sh]: "${WORK_DIR}", please also fix variables below.

https://google.github.io/styleguide/shell.xml?showone=Variable_expansion#Variable_expansion

File: research/deeplab/datasets/build_ade20k_data.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    'Folder containing annotations for validation')
+
+tf.app.flags.DEFINE_string(
+    'output_dir', './ADE20K/tfrecord',
+    'Path to save converted SSTable of Tensorflow example')
+
+_NUM_SHARDS = 4
+
+def _convert_dataset(dataset_split, dataset_dir, dataset_label_dir):
+  """ Convert the ADE20k dataset into into tfrecord format (SSTable).
```
Comment (aquariusjay) [inline research/deeplab/datasets/build_ade20k_data.py]: "" Convert -> """Converts

File: research/deeplab/datasets/build_ade20k_data.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+  Args:
+    dataset_split: dataset split (e.g., train, val)
```
Comment (aquariusjay) [inline research/deeplab/datasets/build_ade20k_data.py]: Make first word capital.

Comment (aquariusjay) [reply research/deeplab/datasets/build_ade20k_data.py]: And put a period (i.e., '.') at the end of each sentence.

File: research/deeplab/datasets/segmentation_dataset.py [hunk 6]
```diff
@@ -85,10 +85,24 @@
...
 )
 
+# These number (i.e., 'train'/'test') seems to have to be hard coded
+# You are required to figure it out for your training/testing example.
+# Is there a way to automatically figure it out ?
+_ADE20K_INFORMATION = DatasetDescriptor(
+    splits_to_sizes = {
+        'train': 20210, # num of samples in images/training
+        'val': 2000, # num of samples in images/validation
+        'eval': 2,
```
Comment (aquariusjay) [inline research/deeplab/datasets/segmentation_dataset.py]: What is this split? Add comment for it? It seems that this split is not generated?

File: research/deeplab/utils/train_utils.py [hunk 7]
```diff
@@ -99,7 +99,7 @@ def get_model_init_fn(train_logdir,
   tf.logging.info('Initializing model from path: %s', tf_initial_checkpoint)
 
   # Variables that will not be restored.
-  exclude_list = ['global_step']
+  exclude_list = ['global_step', 'logits']
```
Comment (aquariusjay) [inline research/deeplab/utils/train_utils.py]: There are cases where you do want to restore `logits' weights (e.g., further fine-tuning them on validation set).
My proposal will be: Could you please

1. Modify get_extra_layer_scopes in model.py as follows.

def get_extra_layer_scopes(last_layers_contain_logits_only=False):
  """Gets the scopes for extra layers.

  Args:
    last_layers_contain_logits_only: Boolean, True if only consider logits
      as the last layers (i.e., exclude ASPP module, decoder module and so on).

  Returns:
    A list of scopes for extra layers.
  """
  if last_layers_contain_logits_only:
    return [_LOGITS_SCOPE_NAME]
  else:
    return [
        _LOGITS_SCOPE_NAME,
        _IMAGE_POOLING_SCOPE,
        _ASPP_SCOPE,
        _CONCAT_PROJECTION_SCOPE,
        _DECODER_SCOPE,
    ]

2. Add a flag in train.py.

flags.DEFINE_boolean('last_layers_contain_logits_only', False,
                     'Only consider logits as last layers or not.')

3. Modify line 295 in train.py as follows.

last_layers = model.get_extra_layer_scopes(
          FLAGS.last_layers_contain_logits_only)

File: research/deeplab/g3doc/ade20k.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+1.  In order to fine tune the BN layers, one needs to use large batch size (> 12),
+    and set fine_tune_batch_norm = True. Here, we simply use small batch size
+    during training for the purpose of demonstration. If the users have limited
+    GPU memory at hand, please fine-tune from our provided checkpoints whose
+    batch norm parameters have been trained, and use smaller learning rate with
+    fine_tune_batch_norm = False.
+
+2. User should fine tune the `min_resize_value` and `max_resize_value` to get
+   better result. Note that `resize_factor` has to equals to `output_stride`.
```
Comment (aquariusjay) [inline research/deeplab/g3doc/ade20k.md]: has to equals -> has to be equal

File: research/deeplab/datasets/segmentation_dataset.py (hunk 6 above)
Comment (walkerlala) [reply research/deeplab/datasets/segmentation_dataset.py]: Oh yes, this split is used for my own testing. I just forgot to remove it. I will turn it into a comment

File: research/deeplab/datasets/build_ade20k_data.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+import string
+import sys
+from PIL import Image
+import build_data
+import tensorflow as tf
+
+FLAGS = tf.app.flags.FLAGS
+flags = tf.app.flags
+
+tf.app.flags.DEFINE_string(
```
[the remark this answers is not in the document]
Comment (walkerlala) [reply research/deeplab/datasets/build_ade20k_data.py]: Ah.. just some ancient code. I will remove it ;-)

File: research/deeplab/datasets/build_ade20k_data.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+# limitations under the License.
+# ==============================================================================
+
+import glob
+import math
+import os
+import random
+import string
+import sys
+from PIL import Image
```
[the remark this answers is not in the document]
Comment (walkerlala) [reply research/deeplab/datasets/build_ade20k_data.py]: This is legacy code. I will remove it.

File: research/deeplab/datasets/build_ade20k_data.py (hunk 2 above)
Comment (walkerlala) [reply research/deeplab/datasets/build_ade20k_data.py]: Thanks!

File: research/deeplab/datasets/download_and_convert_ade20k.sh (hunk 3 above)
Comment (walkerlala) [reply research/deeplab/datasets/download_and_convert_ade20k.sh]: I am just trying to "stay consistent with existing code ;-)". Anyway, I will fix those.

**Comment by walkerlala** [comment]
@YknZhu @aquariusjay All fixed. Please review.

**Comment by RomRoc** [comment]
Hello @walkerlala , could you share a script similar to local_test.sh with all steps to download, convert and retrain a deeplab model with ADE20K dataset, instead of Pascal dataset like in local_test.sh ?
Thanks

**Comment by urgonguyen** [comment]
when I run sh ade_test.sh. It reports the error:

>> Converting image 1/20210 shard 0Traceback (most recent call last):
  File "./build_ade20k_data.py", line 113, in <module>
    tf.app.run()
  File "/home/hangnp/.local/lib/python3.5/site-packages/tensorflow/python/platform/app.py", line 126, in run
    _sys.exit(main(argv))
  File "./build_ade20k_data.py", line 109, in main
    _convert_dataset('train', FLAGS.train_image_folder, FLAGS.train_image_label_folder)
  File "./build_ade20k_data.py", line 92, in _convert_dataset
    image_data = tf.gfile.FastGFile(image_filename, 'r').read()
  File "/home/hangnp/.local/lib/python3.5/site-packages/tensorflow/python/lib/io/file_io.py", line 127, in read
    pywrap_tensorflow.ReadFromStream(self._read_buf, length, status))
  File "/home/hangnp/.local/lib/python3.5/site-packages/tensorflow/python/lib/io/file_io.py", line 95, in _prepare_value
    return compat.as_str_any(val)
  File "/home/hangnp/.local/lib/python3.5/site-packages/tensorflow/python/util/compat.py", line 113, in as_str_any
    return as_str(value)
  File "/home/hangnp/.local/lib/python3.5/site-packages/tensorflow/python/util/compat.py", line 86, in as_text
    return bytes_or_text.decode(encoding)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte

what happen? Help me

**Pull request merged**